// user-service/src/app.js
const express = require('express');
const dotenv = require('dotenv');
const helmet = require('helmet');
const cors = require('cors');
const rateLimit = require('express-rate-limit');

dotenv.config();

const app = express();
const PORT = process.env.PORT || 3000;

// ── Security headers ────────────────────────────────────────────────────
app.use(helmet());

// ── CORS ────────────────────────────────────────────────────────────────
app.use(cors({
  origin: process.env.CORS_ORIGIN || '*',
  methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'],
  allowedHeaders: ['Content-Type', 'Authorization'],
}));

// ── Body parsers ────────────────────────────────────────────────────────
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// ── Rate limiting on auth endpoints ─────────────────────────────────────
const authLimiter = rateLimit({
  windowMs: Number(process.env.RATE_LIMIT_WINDOW_MS) || 15 * 60 * 1000, // 15 minutes
  max: Number(process.env.RATE_LIMIT_MAX) || 20,
  standardHeaders: true,
  legacyHeaders: false,
  message: { message: 'Terlalu banyak percobaan, coba lagi nanti.' },
});

// ── Routes ──────────────────────────────────────────────────────────────
const authRoutes = require('./routes/authRoutes');
const profileRoutes = require('./routes/profileRoutes');
const verifyRoutes = require('./routes/verifyRoutes');

// Health check for Docker HEALTHCHECK & load balancers
app.get('/health', (req, res) => {
  res.json({ status: 'ok', service: 'user-auth', uptime: process.uptime() });
});

app.get('/', (req, res) => {
  res.json({ message: 'TELKANTIN User & Auth Service is running.' });
});

app.use('/api/auth', authLimiter, authRoutes);
app.use('/api/auth', verifyRoutes);
app.use('/api/users', profileRoutes);

// ── Global error handler ────────────────────────────────────────────────
app.use((err, req, res, next) => {
  console.error(err);
  res.status(500).json({ message: 'Internal Server Error' });
});

app.listen(PORT, () => {
  console.log(`User & Auth Service running on port ${PORT}`);
});

module.exports = app;
